core/embed/sec/optiga/optiga.c

```diff
@@ -934,24 +1134,63 @@ optiga_pin_result optiga_pin_verify(
     goto end;
   }
 
-  // Reset the counter which limits the guesses at OID_STRETCHED_PIN.
-  if (optiga_set_data_object(OID_STRETCHED_PIN_CTR, false, COUNTER_RESET,
-                             sizeof(COUNTER_RESET)) != OPTIGA_SUCCESS) {
-    ret = OPTIGA_PIN_ERROR;
-    goto end;
+  ui_progress();
+
+  // Reset the counter which limits the guesses at OID_STRETCHED_PINS.
+  if (pin_index == 0) {
+    if (optiga_reset_counter(OID_STRETCHED_PIN_CTR, PIN_MAX_TRIES) !=
+        OPTIGA_SUCCESS) {
+      ret = OPTIGA_PIN_ERROR;
+      goto end;
+    }
+  } else {
+    // The one extra attempt is used to authorize using the first stretched PIN
+    // in optiga_reset_hmac_counter().
+    if (optiga_reset_counter(OID_STRETCHED_PIN_CTR, PIN_MAX_TRIES + 1) !=
+        OPTIGA_SUCCESS) {
+      ret = OPTIGA_PIN_ERROR;
+      goto end;
+    }
   }
 
   ui_progress();
 
+  optiga_clear_all_auto_states();
+
+  ui_progress();
+
 end:
   memzero(pin_secret, sizeof(pin_secret));
   memzero(digest, sizeof(digest));
-  optiga_clear_auto_state(OID_STRETCHED_PIN);
-  optiga_clear_auto_state(OID_PIN_SECRET);
+  optiga_clear_all_auto_states();
   optiga_set_ui_progress(NULL);
   return ret;
 }
 
+bool optiga_reset_hmac_counter(
+    optiga_ui_progress_t ui_progress,
+    const uint8_t hmac_reset_key[OPTIGA_PIN_SECRET_SIZE]) {
+  ui_progress();
+
+  // Authorize using the first stretched PIN.
+  if (optiga_set_auto_state(OPTIGA_OID_SESSION_CTX, OID_STRETCHED_PINS[0],
+                            hmac_reset_key,
+                            OPTIGA_PIN_SECRET_SIZE) != OPTIGA_SUCCESS) {
+    return false;
+  }
+
+  ui_progress();
+
+  // Reset the counter.
+  if (optiga_reset_counter(OID_PIN_HMAC_CTR, PIN_MAX_TRIES) != OPTIGA_SUCCESS) {
+    return false;
+  }
+
+  ui_progress();
+
+  return true;
```

Comment:
I think it would be good to have a cleanup section here, where we clear all auto states, e.g. if `optiga_reset_counter()` fails, then we leave the auto state open. Not a major security issue I suppose, but not OK.